Framework/PythonInterface/plugins/algorithms/fractional_indexing.py

```diff
@@ -0,0 +1,302 @@
+                     for peak in peaks_workspace])
+
+
+def remove_noninteger(M):
+    """Remove any non integer values from a matrix
+
+    :param M: the matrix to remove non integer values from
+    :return: M with non integer elements set to zero
+    """
+    M[np.absolute(np.mod(M, 1)) > 1e-14] = 0
+    return M
+
+
+def trunc_decimals(vec, n_decimals=2):
+    """Truncate the elements of a vector below n decimals places
+
+    :param vec: the vector to truncate the elements of
+    :param n_decimals: the number of decimal places to truncate from
+    :return: the vector with truncated elements
+    """
+    decade = 10**n_decimals
+    return np.trunc(vec*decade)/decade
+
+
+def sort_vectors_by_norm(vecs):
+    """Sort a list of row vectors by the Euclidean norm
+    :param vecs: vectors to sort according to their norms
+    :returns: ndarray of sorted row vectors
+    """
+    idx = np.argsort(norm_along_axis(np.abs(vecs)))
+    vecs = vecs[idx]
+    return vecs
+
+
+def norm_along_axis(vecs, axis=-1):
+    """Compute the euclidean norm along the rows of a matrix
+
+    :param vecs: vectors to compute the Euclidean norm for
+    :param axis: the axis to compute the norm along
+    :return: a ndarray with the norms along the chosen axis
+    """
+    return np.sum(vecs**2, axis=axis)**(1./2)
+
+
+def kmeans_plus_plus(points, k):
+    """Generate centroids for kmeans using the kmeans++ algorithm
+
+    Select a good set of starting values for kmeans by weighting the choice of
+    the next centroid by the distance to the existing centroids.
+
+    This algorithm roughly works as follows:
+        1) The first centroid is selected uniformly from the data.
+        2) The squared euclidean distance between all points to all centroids is calculated.
+        3) Take the minimum computed distance for each point over all centroids.
+        4) Normalise the list of distances
+        5) Choose a new centroid weighted by the computed distances
+
+    :param points: the points to select centroids from
+    :param k: the desired number of centroids
+    :return: Kx3 matrix of centroids
+    """
+    if points.shape[0] < k:
+        raise RuntimeError("k is greater than the number of points! Please choose a smaller k value")
+
+    centroid_indices = []
+    centroids = []
+
+    # pick the first point uniformly at random from the data
+    indices = np.indices(points.shape)[0][:, 0]
+    centroid_index = np.random.choice(indices)
+    centroid = points[centroid_index]
+
+    centroid_indices.append(centroid_index)
+    centroids.append(centroid)
+
+    for i in range(k-1):
+        # choose all points that are not already centroids
+        mask = np.array([x for x in indices if x not in centroid_indices])
+        pts = points[mask]
+
+        # calculate probability distribution for being picked based on squared
+        # distance from each centroid
+        D2 = np.array([norm_along_axis(pts - c, axis=1)**2 for c in centroids])
```

Comment:
Is `D2` the distance squared?
if so can we change this to pythonic name perhaps `distance2` or the super verbose `distance_squared`